### DISCUSSION TITLE
My opinion: endpoints should be exposed explicitly and in one place (not from different modules)

### DISCUSSION THREAD

**Discussion by andriyndev** [serverpod/serverpod#1001] [General]
That's what I'm going to write may be quite controversial, subjective and "holywarian" but I want to share it anyway to give an opinion of a user who have been using Serverpod for some time for a personal project, and maybe discuss it.
In the current version of Serverpod, I don't like the idea that the modules internally expose endpoints. I think endpoints should be exposed from a single place by user for the sake of security. Endpoints is one of the most crucial parts of server's security, and I, as a Serverpod's user, personally don't feel my server to be safe when it exposes different endpoints under the hood from different places. If I want to review all the endpoint that are exposed by my server, I'll also need to look into each module that is quite cumbersome and easy to miss. Also, the endpoints from another modules cannot be restricted, and upgrading a module to more recent version might implicitly expose additional endpoints that might be undesirable or even dangerous in some cases.
I think the approach should be the same as with routes, i.e. endpoints are exposed from a single place (a single folder with subfolders) within the server while modules provide API that can be accessed by the endpoints.
What others think about the point?

**Comment by Isakdl** [comment]
I think you put the finger on a valid issue, as a project grows it can become hard to know what data is exposed where more so with modules. Ponder that we have many third-party modules, you would really have to review the code to know what data you are potentially exposing. 

With that said adding more configurations to include modules also creates friction especially if this is a requirement to get things to work. Then what if a module needs some specific endpoints enabled to even work? If you can simply toggle things willy-nilly you may end up with some nasty errors, or unpredictable results. These would also have to be maintained over time for example a module adds a new endpoint (required or not) you would have to then manually toggle this on (security wise this is indeed a good thing).

In terms of syntax/file structure, how do you envision such a feature?

**Reply by andriyndev** [reply]
I think modules should provide a set of functions (API) which are called from endpoints implemented by users.

Currently, this would be problematic because the endpoints are called from the front-end part of the module, but, in my personal opinion, Serverpod as a backend framework shouldn't have cared about front-end at all from the beginning. Of course, the current modules make implementing such things like authentication extremely simple by just installing the corresponding module and including the corresponding widget in user's front-end, but, as the consequence, it's impossible to make any changes to the module's backend including restricting endpoints, and the only possible solution is to create a local fork of the module and then manually integrate there changes from the main repository all the time. I think this is a much bigger evil than manually implement endpoints and front-end, and even for small projects it wouldn't be a big problem to implement authentication forms from scratch or reuse existing ones from the Internet. By the way, in my opinion, using a custom front-end instead of front-end from a Serverpod module would always be a good practice, because:
- Freedom to change the design
- Protection from unexpected changing the design after upgrading the module
- Ability to separate view (design) and data (endpoints) layers in the app
- It's more easy to switch to another backend (it might be kind of disadvantage for Serverpod itself but it might also discourage other potential users who consider using Serverpod for their project)
- Ability to implement localization (Serverpod modules can also support this, but the current version of the auth module doesn't)

**Comment by andriyndev** [comment]
Since the issue is quite important (in my humble opinion) and the current approach could be theoretically reconsidered in Serverpod 2.0, it'd be great to have more opinions regarding it (maybe I'm wrong, and the issue is not an issue at all).
I, personally, currently consider switching to another backend because of the issue.
